subprojects/core/src/integTest/groovy/org/gradle/api/tasks/DeferredTaskDefinitionIntegrationTest.groovy

```diff
@@ -825,4 +825,167 @@ class DeferredTaskDefinitionIntegrationTest extends AbstractIntegrationSpec {
         expect:
         succeeds "foo"
     }
+
+    private static final def INVALID_CALL_FROM_LAZY_CONFIGURATION = [
+        ["Project#afterEvaluate(Closure)"  , "afterEvaluate {}"],
+        ["Project#afterEvaluate(Action)"   , "afterEvaluate new Action<Project>() { void execute(Project p) {} }"],
+        ["Project#beforeEvaluate(Closure)" , "beforeEvaluate {}"],
+        ["Project#beforeEvaluate(Action)"  , "beforeEvaluate new Action<Project>() { void execute(Project p) {} }"],
+        ["Project#subprojects(Closure)"    , "subprojects {}"],
+        ["Project#subprojects(Action)"     , "subprojects new Action<Project>() { void execute(Project p) {} }"],
+        ["Project#allprojects(Closure)"    , "allprojects {}"],
+        ["Project#allprojects(Action)"     , "allprojects new Action<Project>() { void execute(Project p) {} }"],
+        ["Project#project(String, Closure)", "project(':nested') {}"],
+        ["Project#project(String, Action)" , "project(':nested', new Action<Project>() { void execute(Project p) {} })"],
+    ]
+
+    @Unroll
+    def "cannot execute #description during lazy task creation action execution"() {
+        settingsFile << "include 'nested'"
+        buildFile << CUSTOM_TASK_WITH_CONSTRUCTOR_ARGS
+        buildFile << """
+            tasks.register("foo") {
+                ${code}
+            }
+        """
+
+        expect:
+        fails "foo"
+        failure.assertHasCause("$description cannot be executed while configuring a lazy task")
```

Comment:
we need a better term than 'lazy task', it's not really the task that is lazy it is the configuration of it that is. Also we should include which task was being configured in the error message, and possibly which project the offending method was called on too.

Reply:
should also mention the breaking change in the release notes.

Reply:
I made the cause more generic and add `@Contextual` to the exception thrown from `CreatingTaskProvider` so the context can be properly read.

I also added a note in the release notes.